File: tmp/tree.c
#include<stdio.h>
#include<stdlib.h>

typedef struct bst
{
    int data;
    struct bst* left;
    struct bst* right;
}bst;

bst *root;

void preorder(bst *root)
{
    if (root != NULL)
    {
        printf("%d ", root->data);
        preorder(root->left);
        preorder(root->right);
    }
}
int insert(bst **node, int data)
{
    if(*node == NULL)
    {
        *node= (bst* )malloc(sizeof(bst));
        if(*node == NULL)
        {
            return -1;
        }
        (*node)->data = data;
        (*node)->left = NULL;
        (*node)->right = NULL;
        return 0;
    }

    if( (*node)->data > data)
    {
        return insert(&(*node)->left, data);
        //return insert(&(*node)->left, data);
        //& and * dont cancel each other.
    }
    else if( (*node)->data < data )
    {
        return insert(&(*node)->right, data);
    }
    else
    {
        printf("Data already present.\n");
        return 0;
    }

}

void print(bst **node) {
        if(*node != NULL)
        {
                printf("%d\n", (*node)->data);
                print(&(*node)->left);
                print(&(*node)->right);
        }
}

int main() {
printf("%d", insert(&root, 2));
printf("%d", insert(&root, 3));
printf("%d", insert(&root, 4));
printf("done:\n");
    preorder(root);
printf("print:\n");
print(&root);
return 1;
}

File: python/ds/tree.py
class node:
    def __init__(self,data):
        self.head=data
        self.left=None
        self.right=None

class tree:
    def __init__(self,data):
        self.head=None
        self.left=None
        self.right=None

    def inorder(self):
        if(self.head):
            self.inorder(self.left)
            print(self.head)
            self.inorder(self.right)

    def insert(self,data):
        if self.head:
            if data < self.head:
                if self.left is None:
                    self.left = tree(data)
                else:
                    self.left.insert(data)
            else:
                if self.right is None:
                    self.right = tree(data)
                else:
                    self.right.insert(data)
        else:
            self.head = tree(data)


if __name__=='__main__':
    ttree = tree(None)
    ttree.insert(1)
    ttree.insert(2)
    ttree.insert(3)
    #ttree.head = node(1)
    #ttree.head.left = node(2)
    #ttree.head.right = node(3)

    ttree.inorder()

